app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt

```diff
@@ -0,0 +1,249 @@
+				Log.e(SpotifyAppController.TAG, "Failed to create the Web API with an authorization code. Authorization code is invalid or not found")
+				createNotAuthorizedNotification()
+				return null
+			}
+
+			return try {
+				val apiBuilder = SpotifyClientApiBuilderHelper.createApiBuilderWithAuthorizationCode(clientId, authorizationCode, spotifyApiOptionsBuilder)
+				apiBuilder.build()
+			} catch (e: SpotifyException.AuthenticationException) {
+				authStateManager.addAuthorizationCodeAuthorizationException(e)
+				Log.e(SpotifyAppController.TAG, "Failed to create the web API with an authorization code. Authorization failed with the error: ${e.message}")
+				createNotAuthorizedNotification()
+				null
+			}
+		}
+	}
+
+	/**
+	 * Updates the [AuthState] with the [Token] information provided.
+	 */
+	private fun updateAuthStateWithAccessToken(token: Token) {
+		val tokenRequest = TokenRequest.Builder(authStateManager.getAuthorizationServiceConfiguration()!!, SpotifyAppController.getClientId(context))
+				.setRefreshToken(token.refreshToken)
+				.setGrantType(GrantTypeValues.REFRESH_TOKEN)
+				.build()
+		val tokenResponse = TokenResponse.Builder(tokenRequest)
+				.setAccessToken(token.accessToken)
+				.setRefreshToken(token.refreshToken)
+				.setAccessTokenExpirationTime(token.expiresAt)
+				.setScopes(token.scopes?.map { it.uri })
+				.build()
+		authStateManager.updateTokenResponse(tokenResponse, null)
+	}
+
+	/**
+	 * Creates and displays a notification stating that the web API needs to be authorized only if the
+	 * user has previously successfully authorized.
+	 */
+	private fun createNotAuthorizedNotification() {
+		if (appSettingsViewer[AppSettings.KEYS.SPOTIFY_SHOW_UNAUTHENTICATED_NOTIFICATION] == "false") {
+			return
+		}
+		val notifyIntent = Intent(context, SpotifyAuthorizationActivity::class.java)
+		notifyIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
+		val notification = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
+				.setContentTitle(context.getString(R.string.notification_title))
+				.setContentText(context.getString(R.string.txt_spotify_auth_notification))
+				.setSmallIcon(R.drawable.ic_notify)
+				.setContentIntent(PendingIntent.getActivity(context, NOTIFICATION_REQ_ID, notifyIntent, PendingIntent.FLAG_UPDATE_CURRENT))
+				.build()
+		val notificationManager = context.getSystemService(NotificationManager::class.java)
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+			val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, NotificationManager.IMPORTANCE_HIGH)
+			notificationManager.createNotificationChannel(channel)
+		}
+		notificationManager.notify(NOTIFICATION_REQ_ID, notification)
+	}
+
+	/**
+	 * Disconnect process for shutting down the [SpotifyClientAPI].
+	 */
+	fun disconnect() {
+		Log.d(TAG, "Disconnecting Web API")
+		webApi?.shutdown()
+		isUsingSpotify = false
+		clearGetLikedSongsAttemptedFlag()
```

Comment:
Clear out the spotifyAppControllerCaller variable, since this is called when the SpotifyAppController disconnects

Reply:
I have it so the clearGetLikedSongsAttemptedFlag() method sets the flag to false and clears out the spotifyAppControllerCaller variable (https://github.com/Ben2356/AndroidAutoIdrive/blob/spotifyWebAPISupport/app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt#L63-L66)

Reply:
You are completely right!